code/cabal-audit/src/Distribution/Audit.hs

```diff
@@ -0,0 +1,246 @@
+  -- ^ path or URL to the advisories
+  , verbosity :: Verbosity.Verbosity
+  -- ^ verbosity of cabal
+  , outputType :: Output
+  }
+
+-- | the main action to invoke
+auditMain :: IO ()
+auditMain =
+  do
+    (auditConfig, nixStyleFlags) <- customExecParser (prefs showHelpOnEmpty) do
+      info (helper <*> auditCommandParser) do
+        mconcat
+          [ fullDesc
+          , progDesc (formatWith [blue] "audit your cabal projects for vulnerabilities")
+          , header (formatWith [bold, blue] "Welcome to cabal audit")
+          ]
+
+    buildAdvisories auditConfig nixStyleFlags
+      >>= handleBuiltAdvisories (outputType auditConfig)
+    `catch` \(SomeException ex) -> do
+      putStrLn $
+        unlines
+          [ formatWith [red, bold] "cabal-audit failed:"
+          , formatWith [red] $ displayException ex
+          ]
+      exitFailure
+
+buildAdvisories :: AuditConfig -> NixStyleFlags () -> IO (M.Map PackageName ElaboratedPackageInfoAdvised)
+buildAdvisories MkAuditConfig {advisoriesPathOrURL, verbosity} flags = do
+  let cliConfig = projectConfigFromFlags flags
+
+  ProjectBaseContext {distDirLayout, cabalDirLayout, projectConfig, localPackages} <-
+    establishProjectBaseContext
+      verbosity
+      cliConfig
+      OtherCommand
+      `catch` \ex -> throwIO $ CabalException {reason = "trying to establish project base context", cabalException = ex}
+  -- the two plans are
+  -- 1. the "improved plan" with packages replaced by in-store packages
+  -- 2. the "original" elaborated plan
+  --
+  -- as far as I can tell, for our use case these should be indistinguishable
+  (_improvedPlan, plan, _, _, _) <-
+    rebuildInstallPlan verbosity distDirLayout cabalDirLayout projectConfig localPackages Nothing
+      `catch` \ex -> throwIO $ CabalException {reason = "elaborating the install-plan", cabalException = ex}
+
+  when (verbosity > Verbosity.normal) do
+    putStrLn (formatWith [blue] "Finished building the cabal install plan, looking for advisories...")
+
+  advisories <- do
+    let k realPath =
+          listAdvisories realPath
+            >>= validation (throwIO . ListAdvisoryValidationError realPath) pure
+    case advisoriesPathOrURL of
+      Left fp -> k fp
+      Right url -> withSystemTempDirectory "cabal-audit" \tmp -> do
+        putStrLn $ formatWith [blue] $ "trying to clone " <> url
+        callProcess "git" ["clone", url, tmp]
+        k tmp
+
+  pure $ matchAdvisoriesForPlan plan advisories
+
+-- | provides the built advisories in some consumable form, e.g. as human readable form
+--
+-- FUTUREWORK(mangoiv): provide output as JSON
+handleBuiltAdvisories :: Output -> M.Map PackageName ElaboratedPackageInfoAdvised -> IO ()
+handleBuiltAdvisories = \case
+  HumanReadable -> humanReadableHandler . M.toList
+  Osv fp -> osvHandler fp
+
+osvHandler :: FilePath -> M.Map PackageName ElaboratedPackageInfoAdvised -> IO ()
+osvHandler fp =
+  Aeson.encodeFile @Value fp . object . M.foldMapWithKey
```

Comment:
What if the parent directory does not exist?

Reply:
If you don't pass a valid `FilePath` that you can write to, you will get an error.

Reply:
or perhaps I misunderstand the issue?

Reply:
I mean, you could try to create the parent directory if it doesn't exist... but that's a design question.

Reply:
I think similar to the `URI` question there's all sorts of things that can go wrong here, for now, I think the errors thrown by the underlying functions will suffice and at some point we can try rewrapping/doing something sensible to circumvent these, if a user happens to need that feature.